compiler/passes/scopeResolve.cpp

```diff
@@ -1954,11 +1994,35 @@ void lookup(const char*           name,
             std::map<Symbol*, VisibilityStmt*>& reexportPts,
             bool storeRenames) {
   llvm::SmallPtrSet<BaseAST*, 32> visited;
+  llvm::SmallPtrSet<Symbol*, 32> foundSymsSet;
 
   lookup(name, context, context, visited, symbols, renameLocs, storeRenames,
-         reexportPts);
+         reexportPts, foundSymsSet);
+}
+
+// Returns true for parenful function; false for parenless fns, vars, types, ...
+static bool isParenfulFn(Symbol* sym) {
+  if (FnSymbol* fn = toFnSymbol(sym)) {
+    return !fn->hasFlag(FLAG_NO_PARENS);
+  }
+
+  return false;
+}
+
+// When traversing up scopes, the scope resolver should stop searching
+// if it finds a variable/type/etc or a parenless function.
+// Continue searching if we have found a method.
+static bool shouldStopSearch(llvm::SmallVectorImpl<Symbol*>& symbols) {
+  for (Symbol* sym : symbols) {
+    if (!isParenfulFn(sym)) {
+      // found parenless fn or type or var or ...
+      return true;
+    }
+  }
+  return false;
 }
```

Comment:
Reading the PR in order (and thus not yet having seen uses for shouldStopSearch) makes me think this is quadratic. After every so often, we will stop and re-run the check on `symbols`? But "every so often" continues as we find more symbols to insert? 

I suppose it might be negligible in the long term (how many overloaded symbols are there really?), but if it becomes a problem I suspect keeping a boolean together with the `symbols` (via some sort of struct, maybe), and updating the flag with each insertion, would be more efficient.

Reply:
Good point. I've added a comment about this strategy, to pursue in the event that loop becomes a performance problem.